kernel/kernel-api/src/main/java/io/delta/kernel/internal/tablefeatures/TableFeatures.java

```diff
@@ -453,31 +466,23 @@ public static Tuple2<Integer, Integer> minimumRequiredVersions(Set<TableFeature>
    * metadata. If the current protocol already satisfies the metadata requirements, return empty.
    *
    * @param newMetadata the new metadata to be applied to the table.
-   * @param needDomainMetadataSupport whether the table needs to explicitly support domain metadata.
-   * @param needClusteringTableFeature whether the table needs to support clustering table feature
-   *     if true it would add domainMetadata support as well.
+   * @param manuallyEnabledFeatures features that were requested to be added to the protocol.
    * @param currentProtocol the current protocol of the table.
    * @return the upgraded protocol and the set of new features that were enabled in the upgrade.
    */
   public static Optional<Tuple2<Protocol, Set<TableFeature>>> autoUpgradeProtocolBasedOnMetadata(
       Metadata newMetadata,
-      boolean needDomainMetadataSupport,
-      boolean needClusteringTableFeature,
+      Collection<TableFeature> manuallyEnabledFeatures,
       Protocol currentProtocol) {
 
     Set<TableFeature> allNeededTableFeatures =
         extractAllNeededTableFeatures(newMetadata, currentProtocol);
-    if (needDomainMetadataSupport) {
+    if (manuallyEnabledFeatures != null && !manuallyEnabledFeatures.isEmpty()) {
       allNeededTableFeatures =
-          Stream.concat(allNeededTableFeatures.stream(), Stream.of(DOMAIN_METADATA_W_FEATURE))
-              .collect(toSet());
-    }
-    // Its dependency feature(domainMetadata) would be enabled automatically.
-    if (needClusteringTableFeature) {
-      allNeededTableFeatures =
-          Stream.concat(allNeededTableFeatures.stream(), Stream.of(CLUSTERING_W_FEATURE))
+          Stream.concat(allNeededTableFeatures.stream(), manuallyEnabledFeatures.stream())
```

Comment:
I think we still need to add dependency features for any of the manuallyEnabledFeatures? Since we currently do this in `extractAllNeededTableFeatures` I think

Reply:
Although I see the test for this is passing. I must be missing something - how are we enabling the dependent features for these?